### PR TITLE
ADD: memory improvements for table models

### PR DESCRIPTION
**Pull request by benleetownsend**
Overall changes drop peak CPU memory usage by ~1GB for a simple test.

### CODE REVIEW

File: finetune/util/table_labeler.py [hunk 1]
```diff
@@ -692,24 +736,21 @@ def fit(
...
     def save(self, path: str) -> None:
-        SequenceLabeler.save_multiple(
-            path,
-            models={
-                "table": self.table_model,
-                "text": self.text_model,
-                "etl": self.etl,
-            },
-        )
+        with open(self.text_model_path, "rb") as text_model, open(
```
Comment (madisonmay) [inline finetune/util/table_labeler.py:740]: Didn't realize you could have multiple context managers on a single line

File: finetune/util/table_labeler.py [hunk 2]
```diff
@@ -677,11 +678,54 @@ def __init__(
...
+    def text_model(self):
+        if os.path.exists(self.text_model_path):
+            text_model = SequenceLabeler.load(self.text_model_path)
+        else:
+            text_model = self._get_text_model()
+        yield text_model
+        text_model.save(self.text_model_path)
+
+    @property
+    @contextlib.contextmanager
```
Comment (madisonmay) [inline finetune/util/table_labeler.py]: Do we need the context manager? As long as we handle all of this within the _fit_text_model function it should go out of scope when the _fit_text_model fn terminates and get cleaned up? NBD obviously, would just be a bit easier to follow flow of code.

**Review by madisonmay — APPROVED** [review]

File: finetune/base.py [hunk 3]
```diff
@@ -649,15 +660,7 @@ def _set_random_seed(self, seed=None):
...
-            for class_name in finetune_model.input_pipeline.label_encoder.target_labels
-            if class_name != self.config.pad_token
-        ]
-
-    @property
-    def classes(self):
-        return [
-            class_name
-            for class_name in finetune_model.input_pipeline.label_encoder.target_labels
+            for class_name in self.input_pipeline.label_encoder.target_labels
```
Comment (madisonmay) [inline finetune/base.py:663]: Oh gross this have must been an artifact of some rebase past.

File: finetune/base_models/bert/table_utils.py [hunk 4]
```diff
@@ -444,7 +444,7 @@ def scatter_feats(output_shape, sequence_feats, scatter_vals):
     """
     input_tensor = tf.zeros(shape=output_shape, dtype=tf.float32)
     mask = tf.math.less(
-        scatter_vals[:, :, 0], output_shape[1]
+        scatter_vals[:, :, 1], output_shape[1]
```
Comment (madisonmay) [inline finetune/base_models/bert/table_utils.py:447]: Having a hard time following shapes here, would you mind documenting expected shape / content of these tensors?

File: finetune/scheduler.py [hunk 5]
```diff
@@ -123,14 +123,17 @@ def _close_oldest_model(self):
...
     def model_cache_key(self, model, key, cache_key):
-        if cache_key and key:
-            return f"{cache_key}_key={key}"
-        elif cache_key and not key:
+        if cache_key is None:
+            if not isinstance(model, str):
+                raise ValueError(
+                    "To schedule a model with a file handle or BytesIO model you must provide a cache_key"
+                )
+            cache_key = f"model={model}"
```
Comment (madisonmay) [inline finetune/scheduler.py:131]: Some of the code related to the cache key is a bit hard to follow because we have two different things called "cache_key" -- one being the raw key provided by the user and the second being the fully formatted cache key that combines the user provided key and cache_key.  It might be worth giving these two different names, or adding a bit of quick validation to make sure a formatted cache_key never gets passed to the model_cache_key function.

File: finetune/util/table_labeler.py [hunk 6]
```diff
@@ -51,14 +56,8 @@ def _adjust_span_to_chunk(
...
-        if output_space_text is not None:
-            assert span["text"] == output_space_text[adj_start:adj_end], (
-                span["text"],
-                output_space_text[adj_start:adj_end],
-                orig_span,
-                ">>",
-                span,
-            )
+        if output_space_text is not None and span["text"] != output_space_text[adj_start:adj_end]:
+            LOGGER.warn("Span Text does not align with output space text")
```
Comment (madisonmay) [inline finetune/util/table_labeler.py:60]: Worth logging some identifying info here that's not sensitive (e.g. start / end?) to aid in debugging?  Though I guess that might not be sufficient to figure out which example was a problem.

File: finetune/util/table_labeler.py [hunk 7]
```diff
@@ -677,39 +677,78 @@ def __init__(
...
+        self._get_text_model = functools.partial(
+            SequenceLabeler, **(text_model_config or {})
+        )
+
+        self.temp_dir = tempfile.TemporaryDirectory()
+        self.text_model_path = os.path.join(self.temp_dir.name, "text.jl")
+        self.table_model_path = os.path.join(self.temp_dir.name, "table.jl")
+        self.classes = set()
+
+    def add_class_names(self, model):
```
Comment (madisonmay) [inline finetune/util/table_labeler.py]: This ensures parity between vanilla finetune model and the table model?

File: finetune/util/table_labeler.py (hunk 6 above)
Comment (benleetownsend) [reply finetune/util/table_labeler.py:60]: Yeah, I couldn't think what might actually be helpful but isn't potentially sensitive.

**Review by madisonmay — APPROVED** [review]
